File: algorithms/16. 3Sum Closest/solution.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        closest = nums[0] + nums[1] + nums[2]
        delta = abs(target - closest)
        for i in range(len(nums)-2):
            j = i + 1
            k = len(nums) - 1
            while j < k:
                s = nums[i] + nums[j] + nums[k]
                if s < target:
                    j += 1
                    if target - s < delta:
                        closest = s
                        delta = target - s
                elif s > target:
                    k -= 1
                    if s - target < delta:
                        closest = s
                        delta = s - target
                else:
                    return s
        return closest

File: algorithms/16. 3Sum Closest/solution.cc
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        long long result = nums[0] + nums[1] + nums[2], tmp;
        int n = nums.size();
        vector<int> snums(nums);
        sort(snums.begin(), snums.end());
        for ( int i, j = 0, k; j < n; ++j ) {
            i = 0;
            k = n - 1;
            while ( i < j && j < k ) {
                tmp = snums[i] + snums[j] + snums[k];
                if ( tmp > target ) {
                    --k;
                } else if ( tmp < target ) {
                    ++i;
                } else {
                    return target;
                }
                if ( abs(tmp - target) < abs(result - target) ) {
                    result = tmp;
                }
            }
        }
        return result;
    }
};
